torch/distributed/distributed_c10d.py

```diff
@@ -2972,18 +2975,61 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
     Returns:
         A handle of distributed group that can be given to collective calls.
     """
+    return _create_new_group_internal(ranks, timeout, backend, pg_options, global_naming=True)
+
+
+def _get_backend_from_str(backend: Optional[str] = None) -> Backend:
+    # Default to the same backend as the global process group
+    # if the backend is not specified.
+    if not backend:
+        default_pg = _get_default_group()
+        default_backend, _ = _pg_map[default_pg]
+        backend = default_backend
+    return Backend(backend)
+
+
+def new_sub_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=None):
```

Comment:
I wonder if this is the best API name, or if we could potentially rename to something indicative of the fact that users don't have to (and can't) call this API on all ranks, only the ones which we're creating the subgroup over.

Reply:
I agree, I'm not sold out on the name either.

But the part of "can't call this on all ranks" is a deliberate choice that's worth exploring a bit.

We could allow calling this on all ranks and those that don't belong have it return None, just line `new_group`, the only difference would be on local vs global synchronization.

The advantage of raising error on a rank that doesn't belong makes it clear that this API behaved differently.
The downside is that migrating to this faster API will be significant more work, I wonder if ppl might end up doing this: 

```
def safe_new_sub_group(ranks):
 return dist.new_sub_group(ranks) if dist.get_rank() in ranks else None
```

If that happens, the current choice is a bad one.

Having said that, c10d is unspecified W.R.T. cluster ordering of operations. `new_group` is no different in that regard. I mean, dist.barrier() has undefined semantics regarding collectives.

What if this was a new optional argument to new_group and had the same behavior when it comes to non-member ranks?

Reply:
After some soul searching, I settled with the following:

- It's a new optional argument `use_local_synchronization` to new_group.
- It returns None if called from a non-member rank.

Making it a parameter makes more sense since it's a slight variation on `new_group` that most people would not notice the difference.

Returning None for non-members makes it significantly easier to transition to this new API as most code can switch to it at no risk of new failures.